dags/trending-projects-snapshot-dag.py

```diff
@@ -0,0 +1,183 @@
+        last_download_date: The date of the last download from the Project
+        total_data_size_in_gib: The current total size of the Project in GiB
+    """
+
+    project_id: float
+    n_unique_users: int
+    last_download_date: int
+    total_data_size_in_gib: float
+
+
+@dag(**dag_config)
+def trending_projects_snapshot() -> None:
+    """
+    This DAG executes a query on Snowflake to retrieve information about trending public projects and 
+    reports the results to a Synapse table.
+
+    The main steps performed in this DAG are:
+    1. Query Snowflake to identify public projects, their file sizes, and recent download activities.
+    2. Aggregate the results to determine num. of unique users, last download date, and total data size.
+    3. Format the results into a list of SnapshotsMetric objects.
+
+    The Snowflake connection and Synapse connection IDs, as well as the time window for recent downloads,
+    are provided via DAG parameters.
+    """
+
+    @task
+    def get_trending_project_snapshot(**context) -> List[SnapshotMetrics]:
+        """Execute the query on Snowflake and return the results."""
+        snow_hook = SnowflakeHook(context["params"]["snowflake_conn_id"])
+        ctx = snow_hook.get_conn()
+        cs = ctx.cursor()
+        query = f"""
+                WITH PUBLIC_PROJECTS AS (
+                    SELECT PROJECT_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST
+                    WHERE 1=1
+                    AND NODE_TYPE = 'project'
+                    AND IS_PUBLIC = TRUE
+                ),
+                LATEST_FILE_HANDLES AS (
+                    SELECT DISTINCT ID, PROJECT_ID, FILE_HANDLE_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST
+                    WHERE 1=1
+                    AND NODE_TYPE = 'file'
+                    AND PROJECT_ID IN (SELECT PROJECT_ID FROM PUBLIC_PROJECTS)
+                ),
+                FILE_SIZES AS (
+                    SELECT fh.PROJECT_ID, SUM(f.CONTENT_SIZE) / POWER(1024, 3) AS TOTAL_DATA_SIZE_IN_GIB
+                    FROM LATEST_FILE_HANDLES fh
+                    JOIN SYNAPSE_DATA_WAREHOUSE.SYNAPSE.FILE_LATEST f
+                    ON fh.FILE_HANDLE_ID = f.ID
+                    GROUP BY fh.PROJECT_ID
+                ),
+                RECENT_DOWNLOADS AS (
+                    SELECT PROJECT_ID, RECORD_DATE, USER_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.FILEDOWNLOAD
+                    WHERE 1=1
+                    AND RECORD_DATE > DATEADD(DAY, -{context["params"]["time_window"]}, CURRENT_DATE()) 
+                    AND STACK = 'prod'
+                    AND PROJECT_ID IN (SELECT PROJECT_ID FROM PUBLIC_PROJECTS)
+                )
+
+                SELECT 
+                    pp.PROJECT_ID,
+                    COUNT(DISTINCT rd.USER_ID) AS N_UNIQUE_USERS,
+                    COALESCE(TO_CHAR(MAX(rd.RECORD_DATE), 'YYYY-MM-DD'), 'N/A') AS LAST_DOWNLOAD_DATE,
```

Comment:
In the event that we have no new downloads within the requested time window for a given project, we `LEFT JOIN` onto `PUBLIC_PROJECTS` and fill `NULL` values with `N/A`.

Reply:
What is your rationale here for shifting this to a character?  the expected output is what we have for "export_date" so keep that in mind.  Nice use of Coalese though for those projects that don't have download dates at all. 

One query optimization is to first get the top 10 downloaded public projects by user, then only grab the data sizes and activity for those projects.

Reply:
> What is your rationale here for shifting this to a character? 

It's because `COALESCE` needs all the arguments to be the same data type (in this case, string). I was getting a `Date 'N/A' is not recognized` error otherwise.

>  the expected output is what we have for "export_date" so keep that in mind

Noted - it looks like the format on Synapse is determined by how the schema is formatted on Synapse itself. I did a test run using the date formatting as it is (`'YYYY-MM-DD'`) but it shows up like `MM/DD/YYYY` on Synapse, same formatting as `export_date`: 
<img width="809" alt="image" src="https://github.com/user-attachments/assets/87742900-6b5e-4a58-88ca-4f489c85c558">

> One query optimization is to first get the top 10 downloaded public projects by user, then only grab the data sizes and activity for those projects.

Nice, thanks! This is what I'm doing with the `LATEST_FILE_HANDLES` CTE where I filter for `PROJECT_ID` in `PUBLIC_PROJECTS`:

```
LATEST_FILE_HANDLES AS (
    SELECT DISTINCT ID, PROJECT_ID, FILE_HANDLE_ID
    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST
    WHERE 1=1
    AND NODE_TYPE = 'file'
    AND PROJECT_ID IN (SELECT PROJECT_ID FROM PUBLIC_PROJECTS)
)
```

and then I also filter `RECENT_DOWNLOADS` with `AND PROJECT_ID IN (SELECT PROJECT_ID FROM PUBLIC_PROJECTS)`

Let me know if you were thinking of something else!

Reply:
> One query optimization is to first get the top 10 downloaded public projects by user, then only grab the data sizes and activity for those projects.

[Updated ✔️ ](https://github.com/Sage-Bionetworks-Workflows/orca-recipes/pull/65/commits/cc55b4e92d33609233282495f50416abe37d11c7)

`RECENT_DOWNLOADS` still works with `PUBLIC_PROJECTS` (not `TOP_10`) because the downloads are what's used to determine what the top 10 projects for that month are. After that, the remaining CTEs use the results from the `TOP_10` project CTE to get the file handles and the corresponding total size for each project, which minimizes calculations. 

Here are my results for the month of June:

<img width="1058" alt="image" src="https://github.com/user-attachments/assets/1cf4c2ff-b8fc-4625-a184-a76cdfaa65e2">

I tried comparing the Query duration between the old and new version of this query, but tbh the calculation time is so small it's hard to tell. I think logically, though, this change does minimize compute.

One other thing that came to mind as I was working on this is we should consider expanding the scope of the trending project snapshots to also include upload activity. In other words `N_UNIQUE_USERS` would include users both uploading and downloading, and we'd have another column for `LAST_UPLOAD_DATE`.